### DISCUSSION TITLE
How to load a single image in MONAI for segmenation?

### DISCUSSION THREAD

**Discussion by Adwaitt** [Project-MONAI/MONAI#3111] [Q&A] [answered]
Hello everyone, hope you are having a good day :)

I wanted to ask how can I load just a single image for segmentation in MONAI. I trained a SegResNet on a nifti image of shape (h, w, number of images = 155) and it stacks the image (4 images) therefore the final output is (4, h, w, number of images = 155) I just want to load only a single image (4, h, w, 1) how can I do that?

Also if I just want to load one single image (no nifti say a png of shape (224, 224)) to meet the requirement of channels is it okay if I just stack the image with just zeros i.e np.dstack((img, np.zeros(h, w), np.zeros(h, w), np.zeros(h, w)))?

Thank you very much for you time :D

**Comment by Nic-Ma** [comment]
Hi @Adwaitt ,

I think the reason it stacked the images is that you passed a list of file paths to `LoadImage` transform, if you only pass 1 path, it will only load 1 image.

Thanks.

**Reply by Adwaitt** [reply]
Thanks for your response Nic. What I meant was, I'm not loading the path via any transform function. The model initially took a nifti volume of size (batch_size, number_of_stacked_images (channels), height, width, number_of_images). Here what I wanna do is, (batch_size, number_of_stacked_images, height width, 1). While I loaded it and fed it to my model it did run but instead of getting 1 output for the segmentation model I got 8 and the mask was not clear at all. (1, 4, 224, 224, 8). Idk why this happened for a single case when I tried it for 144 images the output was (1, 4, 224, 224, 144) and the mask was clear too.

**Comment by Nic-Ma** [comment]
Sorry I don't quite understand your question. Maybe I missed some points, could you please help show your question with some example?
@rijobro @ericspod @wyli Can you guys help share some comments?

Thanks in advance.

**Comment by Adwaitt** [comment, marked as answer]
img = nib.load('path to nib image')
`r = img[:, :, 100]` #Loading the 100th slice from a nifti image of 240 slices
`r = tfms(image = r)['image']` #Applying my transformation
`stacker = np.zeros((224, 224))`
`r = {'image': r, 'label': None}`#dictionary to apply MONAI transforms
`r = new_transform(r)['image']`#New_transform = MONAI transform
`r = np.dstack((r, np.zeros((224, 224)), np.zeros((224, 224)), np.zeros((224, 224))))`# Stacking zeros along depth.`

`r = r[np.newaxis, ...]# Running this code twice to add two new axis so as to make the input compatible with the model`
`r = r.transpose(0, -1, 2, 3, 1) #Transposing the image so as to make it the desired shape`

`with torch.no_grad():`
`    n = model.forward(torch.tensor(r, dtype = torch.float).cuda())`
`    n = np.squeeze(n.cpu().numpy(), 0)`
`n.shape`

Output for a single slice of nifti image (240, 155, 240): (this is what I desire)
![image](https://user-images.githubusercontent.com/67834832/137059253-6d02192b-bb11-44d8-9990-1f4b0ad4f547.png)

Output for single slice of image(240, 155, 1): (this is what I get)
![image](https://user-images.githubusercontent.com/67834832/137059427-3b30bb45-8d68-45fd-bc7b-8fb168a76eb8.png)

Also the output for this single image contains 8 images. Output is (224, 224, 8) (I don't understand why is that)

**Reply by ericspod** [reply]
I would think that the output sice of your network totally depends on its implementation and parameters. What network are you using? What constructor arguments are you using?

**Reply by Adwaitt** [reply]
Sorry for late response Eric, I had power outage in the country for 3 days hence I was unable to respond to you. I am using SegResNet

**Reply by ericspod** [reply]
I think the order of image dimensions that you're passing to the network isn't what you expect. The `SegResNet` class has arguments `in_channels` and `out_channels` in the constructor that stating the number of input and output channels. For an input of shape (B,C,H,W,D) the `in_channels` value must be C, for an output of shape (Bo,Co,Ho,Wo,Do) `out_channels` must be Co. Be sure that your inputs have dimensions ordered (batch, channel, height, width, depth) or some other ordering where batch and channel come first, I'm not sure that's what's going on with your code as you're doing a transpose of dimensions.

**Reply by Adwaitt** [reply]
it is in that order eric. What I am unable to understand (I'm going to write the exact shape) that when I input an image of shape 
(1, 4, 240, 240, 1) I get an output of (8, 240, 240, 1) while the number of labels = 3 (excluding background) what is this 8 and why is it? That is my question. Also thank you for actively responding to my doubts :D to both you and Nic

**Reply by ericspod** [reply]
I'd like to see what your constructor arguments are to SegResNet because that will determine what the network expects. If you want to use the network with 2D slices I think you'd need to specify that, otherwise using it with a depth size of 1 is probably the cause of this issue since it relies on downsampling and upsampling in the spatial dimensions. For example this works as expected:

```python
import torch
from monai.networks.nets import SegResNet

net=SegResNet(
    spatial_dims= 3,
    in_channels = 4,
    out_channels = 3,
    blocks_down = (1, 2, 2, 4),
    blocks_up = (1, 1, 1),
)

t=torch.rand(1, 4, 240, 240, 240)
print(net(t).shape)  # prints torch.Size([1, 3, 240, 240, 240])
```

If I pass in a tensor with a depth size of 1 it will give odd results:

```python
import torch
from monai.networks.nets import SegResNet

net=SegResNet(
    spatial_dims= 3,
    in_channels = 4,
    out_channels = 3,
    blocks_down = (1, 2, 2, 4),
    blocks_up = (1, 1, 1),
)

t=torch.rand(1, 4, 240, 240, 1)
print(net(t).shape)  # prints torch.Size([1, 3, 240, 240, 8])
```

I can see how the 8 in the depth dimension here seems odd but I still don't see how you're getting that in your batch dimension. You can use the network in 2D as such:

```python
import torch
from monai.networks.nets import SegResNet
net=SegResNet(
    spatial_dims= 2,
    in_channels = 4,
    out_channels = 3,
    blocks_down = (1, 2, 2, 4),
    blocks_up = (1, 1, 1),
)

t=torch.rand(1, 4, 240, 240)
print(net(t).shape)
```

**Reply by Adwaitt** [reply]
Eric, you wonderful human being! Thank you, Thank you! It worked. I removed the final 1 and just gave the output as (1, 4, 240, 240) and I got the result! WOW!

**Reply by Adwaitt** [reply]
Eric, I'm unable to "Mark as an answer" button to your reply.

**Reply by Adwaitt** [reply]
Also Thank you Nick and Eric :D MONAI is the best

**Reply by Nic-Ma** [reply]
I marked the thread as the answer.

Thanks.

**Reply by Adwaitt** [reply]
Oh, Thank you Nic and Thank you Team MONAI.

**Reply by Nic-Ma** [reply]
You are welcome.
